ndcube/ndcube.py

```diff
@@ -1330,6 +1330,50 @@ def squeeze(self, axis=None):
         return self[tuple(item)]
 
 
+def fill(self, fill_value, unmask=False, uncertainty_fill_value=None, fill_in_place=False):
+    """
+    Replaces masked data values with input value.
+
+    Returns a new instance or alters values in place.
+
+    Parameters
+    ----------
+    fill_value: `numbers.Number` or scalar `astropy.unit.Quantity`
+        The value to replace masked data with.
+    unmask: `bool`, optional
+        If True, the newly filled masked values are unmasked. If False, they remain masked
+        Default=False
+    uncertainty_fill_value: `numbers.Number` or scalar `astropy.unit.Quantity`, optional
+        The value to replace masked uncertainties with.
+    fill_in_place: `bool`, optional
+        If `True`, the masked values are filled in place.  If `False`, a new instance is returned
+        with masked values filled.  Default=False.
+    """
+    # ...code implementation here.
+    kwargs = {}
+    kwargs["data"] = self.data
+    kwargs["mask"] = self.mask
+    kwargs["uncertainty"] = self.uncertainty
+
+
+    if (fill_in_place is False):
+        # If there is a not None mask and a not None fill_value, do: change the corresponding data to fill_value.
+        if (fill_value is not None and self.mask is not None):
```

Comment:
Given the above assignment of data variables, we no longer need to check `fill_in_place`.

Also, we should support any fill value that's compatible with the data array.  If the user supplies a non-compatible one, the code will automatically error we you try to index the data (or uncertainty) array. So we don't need to check if `fill_value is None``.

However, we can combine the 2nd if condition to check whether there are any masked values.  If there aren't, there's no need to fill them. Then you can produce a new mask that can be used to index the data array (i.e. handle cases where mask is `bool`-type rather than an array).
```suggestion
    masked = False if (self.mask is None or self.mask is False or not self.mask.any()) else True
    if masked:
            idx_mask = slice(None) is self.mask is True else self.mask # Ensure indexing mask can index the data array.
```
Then proceed with filling masked values.

Reply:
Hi @DanRyanIrish, should it be:  
```
if masked:
    idx_mask = slice(None) if self.mask is None or (isinstance(self.mask, np.ndarray) and not np.any(self.mask)) or self.mask is False else self.mask
```
 , instead? 
Meaning, if self.mask is None or all values are False or is a single bool F as a whole (every mask value is T), then there is no need to index the data array at all. Otherwise, use the mask array as the index array?

Reply:
I think your version is unnecessarily long.  Because this line of code is only executed if `masked is True`, then we already know that `self.mask` is not `None` or `False`. Therefore, the only possibilities are that `self.mask` is `True` or a boolean array.  If `self.mask` is `True`, the it can't be used to directly index the data array, therefore we need to set `idx_mask` to a slicing item that captures all elements (i.e. `slice(None)`).  Otherwise, `self.mask` can be used as is.  So I think (but please correct me if I'm wrong) that my suggested line is sufficient.

On a side note inspired by your suggestion of `isinstance(self.mask, np.ndarray)`, `NDCube` does not enforce that all arrays be numpy arrays.  Other arrays exist in Python, e.g. dask arrays.  Therefore, we should never rely on an assumption that arrays are a specific type, e.g. numpy, unless we deliberately and explicitly only want to handle a specific type of array.  And this should be avoided wherever possible.